File: Python/Data_Structures/Linked List/singlylinkedlist.py
# SinglyLinkedList Implementation in Python


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

    def printList(self):
        """
            Printlist Function to Print Nodes in LinkedList manner
        """
        if self.size == 0:
            print("Linked List is Empty !!")
            return

        ptr = self.head
        while ptr:
            print("[ ", ptr.data, " ] -> ", end=" ")
            ptr = ptr.next
        print("NULL")
        print("Length: ", self.size)

    def __len__(self):
        return self.size

    def append_at_top(self, value: int):
        node = Node(value)

        if self.head is None:
            self.head = node
            self.size += 1
            return

        node.next = self.head
        self.head = node
        self.size += 1

    def append_at_bottom(self, value: int):
        node = Node(value)
        ptr = self.head

        if self.size == 0:
            self.head = node
            self.size += 1
            return

        while ptr.next:
            ptr = ptr.next

        ptr.next = node
        self.size += 1

    def append_after(self, previous: int, new: int):
        new_node = Node(new)

        if self.size == 0:
            self.head = new_node
            self.size += 1
            return

        prev_ptr = self.head
        next_ptr = prev_ptr.next
        while prev_ptr.data != previous:
            prev_ptr = next_ptr
            next_ptr = next_ptr.next

        prev_ptr.next = new_node
        new_node.next = next_ptr
        self.size += 1

    def delete_at_top(self):
        if not len(self):
            print("LinkedList is empty !!")
            return

        node = self.head
        self.head = self.head.next
        self.size -= 1

        print("Deleted: ", node.data)

    def reverse_iterative(self):
        prev_ptr = None
        curr_ptr = self.head

        while curr_ptr:
            next_ptr = curr_ptr.next
            curr_ptr.next = prev_ptr
            prev_ptr = curr_ptr
            curr_ptr = next_ptr

        self.head = prev_ptr

    def reverse_recursive(self, prev_ptr, curr_ptr):
        if curr_ptr is None:
            self.head = prev_ptr
            return
        next_ptr = curr_ptr.next
        curr_ptr.next = prev_ptr
        return self.reverse_recursive(curr_ptr, next_ptr)


if __name__ == '__main__':
    sll = SinglyLinkedList()
    sll.append_at_top(12)
    sll.append_at_bottom(45)
    sll.printList()
    sll.append_after(12, 64)
    sll.printList()
    sll.reverse_iterative()
    sll.printList()
    sll.reverse_recursive(None, sll.head)
    sll.printList()
    sll.delete_at_top()
    sll.printList()

File: C++/Data Structures/LinkedList/singlylinkedlist.cpp
//
// @author: Sayantan Paul
//

/*
 * Single LinkedList Implementation in C++
 */


#include <iostream>

// Node Class
class Node {
public:
    int data;       // data inside a node
    Node* next;     // pointer to next node

    // constructor
    Node(int info): data(info), next(NULL) {}

    // Display Node
    void displayNode() {
        std::cout << data << " ";
    }
};

// SinglyLinkedList Class
class SinglyLinkedList {
private:
    Node* head;             // head node of the SinglyLinkedList
    int size = 0;
public:
    // constructor define the head to be NULL
    SinglyLinkedList(): head(NULL) {}

    // destructor -> delete all the nodes of SinglyLinkedList after complete execution of program
    ~SinglyLinkedList() {
        Node* curr = head;
        while (curr != NULL) {
            Node* temp = curr;
            curr = curr->next;
            delete temp;
        }
    }

    // check if singlylinkedlist is empty
    bool isEmpty() {
        return head == NULL;
    }

    // AddToHead -> add node at the top of the singlylinkedlist
    void AddToHead(int data) {
        Node* newNode = new Node(data);
        newNode->next = head;
        head = newNode;
        size++;
    }

    // AddToEnd -> add node at the end of the singlylinkedlist
    void AddToEnd(int data) {
        Node* newNode = new Node(data);
        if (isEmpty()) {
            AddToHead(data);
        } else {
            Node* currNode = head;
            while (currNode->next != NULL) {
                currNode = currNode->next;
            }
            currNode->next = newNode;
            size++;
        }
    }

    // FindData -> Find position of the data in the singlylinkedlist
    int FindData(int data) {
        Node* currNode = head;
        int pos = 0;
        while (currNode != NULL) {
            if (currNode->data == data) {
                std::cout << currNode->data << " found at position " << pos+1 << std::endl;
                return pos;
            }
            currNode = currNode->next;
            pos++;
        }

        std::cout << "[Error]:" << data << " not present" << std::endl;
        return -1;
    }

    // DeleteFromHead -> Delete Node at the top of the singlylinkedlist
    Node* DeleteFromHead() {
        if (head == NULL) {
            std::cout << "[Error]: Linkedlist is empty" << std::endl;
        } else {
            Node* temp = head;
            head = head->next;
            size--;
            return temp;
        }
    }

    // DeleteFromEnd -> Delete Node at the end of the singlylinkedlist
    Node* DeleteFromEnd() {
        if (head == NULL) {
            std::cout << "[Error]: Linkedlist is empty" << std::endl;
        } else {
            Node* prev = head;
            Node* curr = head->next;
            while (curr->next != NULL) {
                prev = curr;
                curr = curr->next;
            }
            prev->next = NULL;
            size--;
            return curr;
        }
    }

    // DeleteData -> Find a data and delete the node
    void DeleteData(int data) {
        Node* prev = head;
        Node* curr = head->next;
        if (prev->data == data) {
            DeleteFromHead();
        } else {
            while (curr != NULL) {
                if (curr->data == data) {
                    curr = curr->next;
                    prev->next = curr;
                    break;
                }
                prev = curr;
                curr = curr->next;
            }
            if (curr == NULL) {
                std::cout << "Data Not Found in SinglyLinkedList" << std::endl;
            }
        }

    }

    // PrintList -> Print out the items in singlylinkedlist
    void PrintList() {
        Node* currNode = head;
        std::cout << "\nLinked List: \n----------------" << std::endl;
        while (currNode != NULL) {
            std::cout << "[ " << currNode->data << " ] --> ";
            currNode = currNode->next;
        }
        std::cout << "NULL\n" << std::endl;
    }

};


int main() {
    SinglyLinkedList* sllist = new SinglyLinkedList();

    sllist->AddToHead(12);
    sllist->AddToEnd(15);
    sllist->PrintList();

    sllist->AddToHead(17);
    sllist->AddToEnd(19);
    sllist->AddToHead(32);
    sllist->AddToEnd(95);
    sllist->PrintList();

    sllist->FindData(32);
    sllist->FindData(17);

    int data = 15;
    sllist->DeleteData(data);
    std::cout << "Deleted Node: " << data << std::endl;
    sllist->PrintList();

    std::cout << "Deleted Node: " << sllist->DeleteFromHead()->data << std::endl;
    std::cout << "Deleted Node: " << sllist->DeleteFromEnd()->data << std::endl;
    sllist->PrintList();


    sllist->FindData(15);
    sllist->FindData(32);

    return 0;
}